File: java/org.financer.client/src/main/java/org/financer/client/connection/HttpMethod.java
package org.financer.client.connection;

import okhttp3.Request;
import okhttp3.RequestBody;
import org.financer.shared.domain.model.api.DataTransferObject;

public interface HttpMethod {

    static HttpMethod fromString(String httpMethod) {
        switch (httpMethod) {
            case "GET":
                return new Get();
            case "POST":
                return new Post();
            case "PUT":
                return new Put();
            case "DELETE":
                return new Delete();
            case "PATCH":
                return new Patch();
            case "OPTIONS":
                return new Options();
            default:
                throw new IllegalArgumentException("Illegal Http Method");
        }
    }

    default RequestBody requestBody(DataTransferObject body) {
        return RequestBody.create(body.toJson(), ServerRequest.MEDIA_TYPE_JSON);
    }

    Request.Builder build(Request.Builder builder, DataTransferObject body);

    class Get implements HttpMethod {
        @Override
        public Request.Builder build(Request.Builder builder, DataTransferObject body) {
            return builder.get();
        }

        @Override
        public String toString() {
            return "GET";
        }
    }

    class Post implements HttpMethod {
        @Override
        public Request.Builder build(Request.Builder builder, DataTransferObject body) {
            return builder.post(requestBody(body));
        }

        @Override
        public String toString() {
            return "POST";
        }
    }

    class Put implements HttpMethod {
        @Override
        public Request.Builder build(Request.Builder builder, DataTransferObject body) {
            return builder.post(requestBody(body));
        }

        @Override
        public String toString() {
            return "PUT";
        }
    }

    class Delete implements HttpMethod {
        @Override
        public Request.Builder build(Request.Builder builder, DataTransferObject body) {
            return builder.delete();
        }

        @Override
        public String toString() {
            return "DELETE";
        }
    }

    class Patch implements HttpMethod {
        @Override
        public Request.Builder build(Request.Builder builder, DataTransferObject body) {
            return builder.patch(requestBody(body));
        }

        @Override
        public String toString() {
            return "PATCH";
        }
    }

    class Options implements HttpMethod {
        @Override
        public Request.Builder build(Request.Builder builder, DataTransferObject body) {
            return builder.method("OPTIONS", null);
        }

        @Override
        public String toString() {
            return "OPTIONS";
        }
    }
}

File: java/org.financer.client/src/main/kotlin/org/financer/client/connection/HttpMethod.kt
package org.financer.client.connection

import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.financer.shared.domain.model.api.DataTransferObject

interface HttpMethod {
    fun requestBody(body: DataTransferObject): RequestBody? {
        return body.toJson().toRequestBody(ServerRequest.MEDIA_TYPE_JSON)
    }

    fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder
    class Get : HttpMethod {
        override fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder {
            return builder.get()
        }

        override fun toString(): String {
            return "GET"
        }
    }

    class Post : HttpMethod {
        override fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder {
            return builder.post(requestBody(body!!)!!)
        }

        override fun toString(): String {
            return "POST"
        }
    }

    class Put : HttpMethod {
        override fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder {
            return builder.post(requestBody(body!!)!!)
        }

        override fun toString(): String {
            return "PUT"
        }
    }

    class Delete : HttpMethod {
        override fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder {
            return builder.delete()
        }

        override fun toString(): String {
            return "DELETE"
        }
    }

    class Patch : HttpMethod {
        override fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder {
            return builder.patch(requestBody(body!!)!!)
        }

        override fun toString(): String {
            return "PATCH"
        }
    }

    class Options : HttpMethod {
        override fun build(builder: Request.Builder, body: DataTransferObject?): Request.Builder {
            return builder.method("OPTIONS", null)
        }

        override fun toString(): String {
            return "OPTIONS"
        }
    }

    companion object {
        fun fromString(httpMethod: String?): HttpMethod {
            return when (httpMethod) {
                "GET" -> Get()
                "POST" -> Post()
                "PUT" -> Put()
                "DELETE" -> Delete()
                "PATCH" -> Patch()
                "OPTIONS" -> Options()
                else -> throw IllegalArgumentException("Illegal Http Method")
            }
        }
    }
}